File: scripts/llm_utils.js
/**
 * LLM Utilities for AI Automation
 * Battle-tested helpers for working with OpenAI, Anthropic, and local models.
 * 
 * @module llm_utils
 */

/**
 * @typedef {Object} LLMConfig
 * @property {'openai'|'anthropic'|'local'} provider
 * @property {string} model
 * @property {number} maxTokens
 * @property {number} temperature
 * @property {number} timeout
 * @property {number} maxRetries
 * @property {number} retryDelay
 */

/** @type {LLMConfig} */
const DEFAULT_CONFIG = {
  provider: 'openai',
  model: 'gpt-4o-mini',
  maxTokens: 1000,
  temperature: 0.1, // Low for automation tasks
  timeout: 30000,
  maxRetries: 3,
  retryDelay: 1000,
};

/**
 * Sleep helper
 * @param {number} ms 
 */
const sleep = (ms) => new Promise(resolve => setTimeout(resolve, ms));

/**
 * Retry wrapper with exponential backoff
 * @template T
 * @param {() => Promise<T>} fn 
 * @param {number} maxRetries 
 * @param {number} baseDelay 
 * @returns {Promise<T>}
 */
async function retryWithBackoff(fn, maxRetries = 3, baseDelay = 1000) {
  let lastError;
  
  for (let attempt = 0; attempt < maxRetries; attempt++) {
    try {
      return await fn();
    } catch (err) {
      lastError = err;
      if (attempt < maxRetries - 1) {
        const delay = baseDelay * Math.pow(2, attempt);
        await sleep(delay);
      }
    }
  }
  
  throw lastError;
}

/**
 * Unified client for multiple LLM providers
 */
class LLMClient {
  /**
   * @param {Partial<LLMConfig>} config 
   */
  constructor(config = {}) {
    this.config = { ...DEFAULT_CONFIG, ...config };
    this._client = null;
  }

  /**
   * Initialize the appropriate client
   * @private
   */
  async _initClient() {
    if (this._client) return this._client;

    if (this.config.provider === 'openai') {
      const { default: OpenAI } = await import('openai');
      this._client = new OpenAI();
    } else if (this.config.provider === 'anthropic') {
      const { default: Anthropic } = await import('@anthropic-ai/sdk');
      this._client = new Anthropic();
    } else {
      throw new Error(`Unknown provider: ${this.config.provider}`);
    }

    return this._client;
  }

  /**
   * Get completion from LLM
   * @param {Object} options
   * @param {string} options.prompt - User prompt
   * @param {string} [options.system] - System prompt
   * @param {boolean} [options.jsonMode] - Parse response as JSON
   * @returns {Promise<Object|string>}
   */
  async complete({ prompt, system = '', jsonMode = true }) {
    return retryWithBackoff(async () => {
      await this._initClient();

      if (this.config.provider === 'openai') {
        return this._openaiComplete(prompt, system, jsonMode);
      } else if (this.config.provider === 'anthropic') {
        return this._anthropicComplete(prompt, system, jsonMode);
      }
    }, this.config.maxRetries, this.config.retryDelay);
  }

  /**
   * @private
   */
  async _openaiComplete(prompt, system, jsonMode) {
    const messages = [];
    if (system) {
      messages.push({ role: 'system', content: system });
    }
    messages.push({ role: 'user', content: prompt });

    const params = {
      model: this.config.model,
      messages,
      max_tokens: this.config.maxTokens,
      temperature: this.config.temperature,
    };

    if (jsonMode) {
      params.response_format = { type: 'json_object' };
    }

    const response = await this._client.chat.completions.create(params);
    const content = response.choices[0].message.content;

    return jsonMode ? JSON.parse(content) : content;
  }

  /**
   * @private
   */
  async _anthropicComplete(prompt, system, jsonMode) {
    const params = {
      model: this.config.model,
      max_tokens: this.config.maxTokens,
      messages: [{ role: 'user', content: prompt }],
    };

    if (system) {
      params.system = system;
    }

    const response = await this._client.messages.create(params);
    const content = response.content[0].text;

    return jsonMode ? this._extractJson(content) : content;
  }

  /**
   * Extract JSON from text that might have markdown or other formatting
   * @private
   * @param {string} text 
   * @returns {Object}
   */
  _extractJson(text) {
    // Try direct parse first
    try {
      return JSON.parse(text);
    } catch {}

    // Try to find JSON in markdown code blocks
    const codeBlockMatch = text.match(/```(?:json)?\s*(\{[\s\S]*?\})\s*```/);
    if (codeBlockMatch) {
      return JSON.parse(codeBlockMatch[1]);
    }

    // Try to find raw JSON object
    const jsonMatch = text.match(/\{[\s\S]*\}/);
    if (jsonMatch) {
      return JSON.parse(jsonMatch[0]);
    }

    throw new Error(`Could not extract JSON from response: ${text.slice(0, 200)}`);
  }
}

/**
 * Render a prompt template with variables
 * Supports {{variable}} syntax
 * @param {string} template 
 * @param {Record<string, any>} variables 
 * @returns {string}
 */
function renderPrompt(template, variables) {
  let result = template;
  
  for (const [key, value] of Object.entries(variables)) {
    const placeholder = `{{${key}}}`;
    const stringValue = typeof value === 'object' 
      ? JSON.stringify(value, null, 2) 
      : String(value);
    result = result.replaceAll(placeholder, stringValue);
  }
  
  return result;
}

/**
 * Process items in batches with rate limiting
 * @template T, R
 * @param {T[]} items 
 * @param {(item: T) => Promise<R>} processor 
 * @param {Object} [options]
 * @param {number} [options.batchSize]
 * @param {number} [options.delayBetweenBatches]
 * @returns {Promise<R[]>}
 */
async function batchProcess(items, processor, { batchSize = 10, delayBetweenBatches = 1000 } = {}) {
  const results = [];
  
  for (let i = 0; i < items.length; i += batchSize) {
    const batch = items.slice(i, i + batchSize);
    const batchResults = await Promise.all(batch.map(processor));
    results.push(...batchResults);
    
    if (i + batchSize < items.length) {
      await sleep(delayBetweenBatches);
    }
  }
  
  return results;
}

// ES Module exports
export { LLMClient, renderPrompt, batchProcess, retryWithBackoff, DEFAULT_CONFIG };

// Example usage
if (import.meta.url === `file://${process.argv[1]}`) {
  const client = new LLMClient({
    provider: 'openai',
    model: 'gpt-4o-mini',
    temperature: 0.1,
  });

  const result = await client.complete({
    prompt: "Classify this email: 'Hi, I'd like to upgrade my plan.'",
    system: "You are an email classifier. Return JSON with 'category' and 'priority' fields.",
    jsonMode: true,
  });

  console.log(JSON.stringify(result, null, 2));
}

File: scripts/llm_utils.py
"""
LLM Utilities for AI Automation
Battle-tested helpers for working with OpenAI, Anthropic, and local models.
"""

import json
import time
from typing import Optional, Dict, Any, List
from dataclasses import dataclass
from functools import wraps


@dataclass
class LLMConfig:
    """Configuration for LLM calls"""
    provider: str = "openai"  # openai, anthropic, local
    model: str = "gpt-4o-mini"
    max_tokens: int = 1000
    temperature: float = 0.1  # Low for automation tasks
    timeout: int = 30
    max_retries: int = 3
    retry_delay: float = 1.0


def retry_with_backoff(max_retries: int = 3, base_delay: float = 1.0):
    """Decorator for retrying failed API calls with exponential backoff"""
    def decorator(func):
        @wraps(func)
        def wrapper(*args, **kwargs):
            last_exception = None
            for attempt in range(max_retries):
                try:
                    return func(*args, **kwargs)
                except Exception as e:
                    last_exception = e
                    if attempt < max_retries - 1:
                        delay = base_delay * (2 ** attempt)
                        time.sleep(delay)
            raise last_exception
        return wrapper
    return decorator


class LLMClient:
    """Unified client for multiple LLM providers"""
    
    def __init__(self, config: LLMConfig):
        self.config = config
        self._client = self._init_client()
    
    def _init_client(self):
        if self.config.provider == "openai":
            from openai import OpenAI
            return OpenAI()
        elif self.config.provider == "anthropic":
            from anthropic import Anthropic
            return Anthropic()
        else:
            raise ValueError(f"Unknown provider: {self.config.provider}")
    
    @retry_with_backoff(max_retries=3)
    def complete(
        self,
        prompt: str,
        system: str = "",
        json_mode: bool = True
    ) -> Dict[str, Any]:
        """
        Get completion from LLM.
        Returns parsed JSON if json_mode=True, raw text otherwise.
        """
        if self.config.provider == "openai":
            return self._openai_complete(prompt, system, json_mode)
        elif self.config.provider == "anthropic":
            return self._anthropic_complete(prompt, system, json_mode)
    
    def _openai_complete(self, prompt: str, system: str, json_mode: bool):
        messages = []
        if system:
            messages.append({"role": "system", "content": system})
        messages.append({"role": "user", "content": prompt})
        
        kwargs = {
            "model": self.config.model,
            "messages": messages,
            "max_tokens": self.config.max_tokens,
            "temperature": self.config.temperature,
        }
        
        if json_mode:
            kwargs["response_format"] = {"type": "json_object"}
        
        response = self._client.chat.completions.create(**kwargs)
        content = response.choices[0].message.content
        
        if json_mode:
            return json.loads(content)
        return content
    
    def _anthropic_complete(self, prompt: str, system: str, json_mode: bool):
        kwargs = {
            "model": self.config.model,
            "max_tokens": self.config.max_tokens,
            "messages": [{"role": "user", "content": prompt}],
        }
        
        if system:
            kwargs["system"] = system
        
        response = self._client.messages.create(**kwargs)
        content = response.content[0].text
        
        if json_mode:
            # Extract JSON from response
            return self._extract_json(content)
        return content
    
    @staticmethod
    def _extract_json(text: str) -> Dict[str, Any]:
        """Extract JSON from text that might have markdown or other formatting"""
        # Try direct parse first
        try:
            return json.loads(text)
        except json.JSONDecodeError:
            pass
        
        # Try to find JSON in markdown code blocks
        import re
        json_match = re.search(r'```(?:json)?\s*(\{.*?\})\s*```', text, re.DOTALL)
        if json_match:
            return json.loads(json_match.group(1))
        
        # Try to find raw JSON object
        json_match = re.search(r'\{.*\}', text, re.DOTALL)
        if json_match:
            return json.loads(json_match.group(0))
        
        raise ValueError(f"Could not extract JSON from response: {text[:200]}")


def render_prompt(template: str, variables: Dict[str, Any]) -> str:
    """
    Render a prompt template with variables.
    Supports {{variable}} syntax.
    """
    result = template
    for key, value in variables.items():
        placeholder = "{{" + key + "}}"
        if isinstance(value, (dict, list)):
            value = json.dumps(value, indent=2)
        result = result.replace(placeholder, str(value))
    return result


def batch_process(
    items: List[Any],
    processor: callable,
    batch_size: int = 10,
    delay_between_batches: float = 1.0
) -> List[Any]:
    """
    Process items in batches with rate limiting.
    Useful for staying within API limits.
    """
    results = []
    for i in range(0, len(items), batch_size):
        batch = items[i:i + batch_size]
        batch_results = [processor(item) for item in batch]
        results.extend(batch_results)
        
        if i + batch_size < len(items):
            time.sleep(delay_between_batches)
    
    return results


# Example usage
if __name__ == "__main__":
    config = LLMConfig(
        provider="openai",
        model="gpt-4o-mini",
        temperature=0.1
    )
    
    client = LLMClient(config)
    
    result = client.complete(
        prompt="Classify this email: 'Hi, I'd like to upgrade my plan.'",
        system="You are an email classifier. Return JSON with 'category' and 'priority' fields.",
        json_mode=True
    )
    
    print(json.dumps(result, indent=2))
